lib/experimental/fonts/class-wp-font-face-resolver.php

```diff
@@ -0,0 +1,158 @@
+<?php
+/**
+ * WP_Font_Face_Resolver class.
+ *
+ * @package    WordPress
+ * @subpackage Fonts
+ * @since      X.X.X
+ */
+
+if ( class_exists( 'WP_Font_Face_Resolver' ) ) {
+	return;
+}
+
+/**
+ * The Font Face Resolver abstracts the processing of different data sources
+ * (such as theme.json) for processing within the Font Face.
+ *
+ * This class is for internal core usage and is not supposed to be used by
+ * extenders (plugins and/or themes).
+ *
+ * @access private
+ */
+class WP_Font_Face_Resolver {
+
+	/**
+	 * Gets fonts defined in theme.json.
+	 *
+	 * @since X.X.X
+	 *
+	 * @return array Returns the font-families, each with their font-face variations.
+	 */
+	public static function get_fonts_from_theme_json() {
+		$settings = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_settings();
+
+		// Bail out early if there are no font settings.
+		if ( empty( $settings['typography'] ) || empty( $settings['typography']['fontFamilies'] ) ) {
+			return array();
+		}
+
+		return static::parse_settings( $settings );
+	}
+
+	/**
+	 * Parse theme.json settings to extract font definitions with variations grouped by font-family.
+	 *
+	 * @since X.X.X
+	 *
+	 * @param array $settings Font settings to parse.
+	 * @return array Returns an array of fonts, grouped by font-family.
+	 */
+	private static function parse_settings( array $settings ) {
+		$fonts = array();
+
+		foreach ( $settings['typography']['fontFamilies'] as $font_families ) {
+			foreach ( $font_families as $definition ) {
+
+				// Skip if font-family "name" is not defined.
+				if ( empty( $definition['name'] ) ) {
+					continue;
+				}
+
+				// Skip if "fontFace" is not defined, meaning there are no variations.
+				if ( empty( $definition['fontFace'] ) ) {
+					continue;
+				}
+
+				$font_family = $definition['name'];
+
+				// Prepare the fonts array structure for this font-family.
+				if ( ! array_key_exists( $font_family, $fonts ) ) {
+					$fonts[ $font_family ] = array();
+				}
+
+				$fonts[ $font_family ] = static::convert_font_face_properties( $definition['fontFace'], $font_family );
+			}
+		}
+
+		return $fonts;
+	}
+
+	/**
+	 * Converts font-face properties from theme.json format.
+	 *
+	 * @since X.X.X
+	 *
+	 * @param array  $font_face_definition The font-face definitions to convert.
+	 * @param string $font_family_property The value to store in the font-face font-family property.
+	 * @return array Converted font-face properties.
+	 */
+	private static function convert_font_face_properties( array $font_face_definition, $font_family_property ) {
+		$converted_font_faces = array();
+
+		foreach ( $font_face_definition as $font_face ) {
+			// Add the font-family property to the font-face.
+			$font_face['font-family'] = $font_family_property;
+
+			// Converts the "file:./" src placeholder into a theme font file URI.
+			if ( ! empty( $font_face['src'] ) ) {
+				$font_face['src'] = static::to_theme_file_uri( (array) $font_face['src'] );
```

Comment:
I know this is in Core already, but it relates to a feature being built for 6.6: [relative paths to background images in theme.json styles](https://github.com/WordPress/gutenberg/pull/61271)

For consistency, and after a nudge from @creativecoder, I decided to follow the `file:./` dotslash path convention to represent paths to these images.

@noisysocks then pointed out that dotslash indicates the current directory, and might imply other dot-style paths, e.g., `../`

So, if web fonts can be used in theme variations, e.g., `my-theme/styles/variation-with-fonts.json`, the `./` would indicate a path relative to the styles directory.

It might also imply that paths with `../../some/path.file` could also be resolved. But I'm not sure if `get_theme_file_uri()` supports such things as it looks for files relative to the active or parent theme root. Maybe I'm missing something.

https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.4/fonts/font-face/class-wp-font-face-resolver.php#L147C27-L147C44

I was wondering if we should enforce a convention. Either:

1. Enforce that all paths must be relative to the theme root, or
2. Allow relative to current directory, and then in resolve them in the background with PHP trickery with [realpath](https://www.php.net/manual/en/function.realpath.php) or something.

What do folks think?

Reply:
I see what you mean... the syntax and behavior seem mismatched: `file:./` seems to imply a reference to the current directory, but really it's a placeholder for the theme root looking at how it's coded.

Updating `file:./` to apply to the current directory (e.g. `theme-root/styles`) for theme variations and/or adding resolution of `../` paths would make the syntax clearer. But I worry about backward compatibility (I imagine that some theme variations are already using `file:./`?). And some care is needed to prevent directory traversal outside the theme directory.

So I favor keeping things simple, and continuing to use `file:./` simply as a placeholder to the theme directory root. But it's not a strong opinion.

(Note that either way, we should add to and update the [theme handbook docs](https://developer.wordpress.org/themes/global-settings-and-styles/settings/typography/#registering-web-fonts-font-faces) to make it clearer).

Reply:
Thanks for the quick reply @creativecoder 

> So I favor keeping things simple, and continuing to use file:./ simply as a placeholder to the theme directory root. 

👍🏻 

I went with this approach for now given that folks might be using to it already. 

> Updating file:./ to apply to the current directory (e.g. theme-root/styles) for theme variations and/or adding resolution of ../ paths would make the syntax clearer. But I worry about backward compatibility (I imagine that some theme variations are already using file:./?).

True. Backwards compatibility is now a requirement regardless. And I considered this in https://github.com/WordPress/gutenberg/pull/61271 and contemplated supporting `file:./` (with the `./` for folks who are now used to it) and also a `file:` prefix. 

But I agree that it's simpler to go with what's there until folks come up with a viable alternative. Even if it's _"Enforce that all paths must be relative to the theme root"_ - that way, we could update the docs now and make the `./` optional later without adding any complex `../` support etc. 🤷🏻